nhost/compose/ports.go

```diff
@@ -26,3 +63,116 @@ func NewPorts(proxy uint32) Ports {
 	p[SvcTraefik] = proxy
 	return p
 }
+
+// hasura console exposes 2 ports, one for the console itself and one for the migrate api
+func (p Ports) isHasuraConsole(svc string) bool {
+	return svc == HasuraConsole || svc == HasuraConsoleApiPort
+}
+
+// EnsurePortsAvailable ensures that the ports are available and returns a new instance of Ports to prevent unexpected mutations
+func (p Ports) EnsurePortsAvailable(ctx context.Context, config *Config) (Ports, error) {
```

Comment:
I'd suggest making this method private and calling it directly inside `NewPorts`. Both methods seem to go hand in hand in the code so it may simplify the API usage

Reply:
I thought about it, but I'm making a call to docker-compose to get the list of running services. Therefore, I need to have a `*compose.Config` thing to pass and `*compose.Config` uses `Ports` itself.

Reply:
check it out - https://github.com/nhost/cli/blob/3859ae74b9272e1e44503113d911a4f064532d92/nhost/service/manager.go#L47-L50

Reply:
I see two problems there:

1. You have a circular dependency (ports -> dockerCompose -> ports)
2. dockerCompose seems to hold a reference of p which is not the same as `ports` which I feel is not what we want/need (maybe the reference in `dockerCompose` is never used or maybe it is overwritten later?)

The circular dependency isn't great though.

Reply:
It's a cirtular dependency indeed. `compose.Config` holds a reference of `p` which is the same as `Ports`. `compose.Config` generates a config for `docker compose` and therefore needs information about ports to expose for certain services. On the flip side we need to query docker compose to get the list of running services to prevent ports collisions.